### DISCUSSION TITLE
No Mesh Local address on Wifi network

### DISCUSSION THREAD

**Discussion by OlivierGre** [openthread/openthread#7698] [Q&A] [answered]
Hi, 
I have a question regarding IPV6 addresses needed to use Thread commissioning.

If I use my Home wifi connection, I can see that my smartphone receives IPV4 addresses and Link Local IPV6 addresses (starting with fe80). As per my understanding of https://openthread.io/guides/thread-primer/ipv6-addressing, these Link Local IPV6 addresses are not suitable for a communication between the smartphone (on wifi network) and a Thread device (on Thread network) because such Link Local address will not be routed by the Open Thread Border Router. A packet sent on Thread network by a Thread device will not be able to reach a smartphone which is on Wifi network. To achieve this, I need a Mesh Local or global address.

I have an alternate solution using my cell phone. With that solution, I use 2 smartphones: 
- One smartphone (connected to 4G) is set as a Wifi access point.
- One smartphone is used for the commissioning. It is connected to the Wifi network created by the other smartphone.
In that case, the smartphone used for the commissioning receives "Mesh Local" IPV6 addresses, which is fine for routing packets between Wifi network and Thread network. This is the setup that I'm using every days. I can do the commissioning and I can then send packets between the smartphone and the Thread device.

How are we supposed to handle the first case where the Wifi network only gives Link Local IPV6 addresses? 

Thank you!

**Comment by jwhui** [comment, marked as answer]
An IPv6 router on the network needs to advertise a routable IPv6 prefix that allows Wi-Fi and Thread devices to communicate with each other. The default OpenThread Border Router configuration should automatically detect if there are no existing IPv6 routers on the Wi-Fi network and start advertising its own routable IPv6 prefix on the Wi-Fi network. The [Thread Border Router - Bidirectional IPv6 Connectivity and DNS-Based Service Discovery](https://openthread.io/codelabs/openthread-border-router) guide should demonstrate this behavior.

If you use tcpdump to capture IPv6 packets on the Wi-Fi network, you should see ICMPv6 Router Advertisement messages being sent on the WI-Fi network.

**Reply by OlivierGre** [reply]
Thank you very much Jonathan!

I have put below a drawing showing my setup. 
Here are the netdata of my Border Router, with its routable IPV6 prefix:
```
netdata show 
Prefixes: 
fd2b:8d62:4ff8:a5dc::/64 paos med 3000 
Routes: 
fd84:c4cb:75d8:0::/64 s med 3000 
Services: 
44970 5d fd111111112200009b9a133d7bda5cf0d11f s 3000 
Done 
```

If my understanding is correct, the smartphone can discover the Thread device (if an advertising is done through SRP), get its IPV6 address (starting with the "routable IPV6 prefix") and send some datagrams to the Thread Device.
But what about the opposite? Is the Thread device able to send some IPV6 packets to the smartphone? I don't see how the smartphone would get a routable IPV6 address.

Thank you

<img width="827" alt="network" src="https://user-images.githubusercontent.com/34505085/168548084-ee758207-2ebc-4b3f-ab7c-4df9b92fce2a.png">

**Reply by wgtdkp** [reply]
> But what about the opposite? Is the Thread device able to send some IPV6 packets to the smartphone? I don't see how the smartphone would get a routable IPV6 address.

Yes, you can find the routable IPv6 address of the smartphone with the Network Analyzer App mentioned in https://openthread.io/codelabs/openthread-border-router#6. Then can ping the smartphone directly from the end device with the OT CLI `ping` command.

**Reply by OlivierGre** [reply]
As per my first message, at the top of this thread, what you say works fine when I use a smartphone configured in Wifi Access Point (receiving data from 4G network). 
However, if I use my Home Wifi access point, then the Commissioner only receives  IPV4 addresses and Link Local IPV6 addresses (starting with fe80). This is in this case that I don't see how the Thread device could send a datagram to the Commissioner.

**Reply by wgtdkp** [reply]
A Thread Border Router will send IPv6 Router Advertisement messages to config routable IPv6 addresses on all Wi-Fi devices in the same Wi-Fi network if it finds out that there is no IPv6 service provided in the Wi-Fi network. It means once you set up a Thread Border Router in your Wi-Fi network, there will always be routable IPv6 addresses on your smartphone.

See https://datatracker.ietf.org/doc/html/rfc4862 and https://datatracker.ietf.org/doc/html/rfc4861 for backgrounds how the "routable" addresses are being configured.

**Reply by jwhui** [reply]
> If my understanding is correct, the smartphone can discover the Thread device (if an advertising is done through SRP), get its IPV6 address (starting with the "routable IPV6 prefix") and send some datagrams to the Thread Device.
But what about the opposite? Is the Thread device able to send some IPV6 packets to the smartphone?

Note that if your smartphone can send packets to the Thread device, the smartphone already has a routable IPv6 address (which it includes in the IPv6 Source Address field of packets sent to the Thread device).

> I don't see how the smartphone would get a routable IPV6 address.

As mentioned in https://github.com/openthread/openthread/discussions/7698#discussioncomment-2745982, the border router sends ICMPv6 Router Advertisement messages with a Prefix Information Option allowing the smartphone to self-configure a routable IPv6 address.

**Reply by OlivierGre** [reply]
Thank you very much!
